### PR TITLE
Run pod-creating controllers under infra-namespace service accounts

### LINKED ISSUE #2774: Uniquely identify BuildPodController user
Currently builders are run under the user `system:openshift-client`.  However, builders may require escalated privileges which need to be controller by a security context constraint.  Granting an SCC to `system:openshift-client` is not a good solution since replication controllers also run with that user and we would effectively give escalated privileges to all pods.

/cc @csrwng @liggitt @pmorie

### PR DESCRIPTION
**Pull request by liggitt**
- [x] Add config option for infra namespace
- [x] Ensure namespace exists on startup
- [x] Ensure replicationController, buildController, deploymentController service accounts exist
- [x] Define roles for ^
- [x] Ensure service accounts are mapped to roles at startup
- [x] Build clients using service account credentials when starting controllers
- [ ] Open issue for doc to add infra namespace config
- [ ] Open issue for install to add infra namespace config to templates
- [x] check this item

Resolves #2774

### CODE REVIEW

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
@smarterclayton @pweil- still working on tests, but ready for review

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 1]
```diff
@@ -213,6 +213,100 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
 		},
 		{
 			ObjectMeta: kapi.ObjectMeta{
+				Name: BuildControllerRoleName,
```
Comment (liggitt) [inline pkg/cmd/server/bootstrappolicy/policy.go]: @bparees can you take a look at this list of permissions, and let me know if you see anything else the controllers launched from RunBuildController() need? Note that RunBuildPodController and RunBuildImageChangeTriggerController are still using the highly-privileged client.

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 2]
```diff
@@ -213,6 +213,100 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
...
+				// BuildController.Recorder (EventBroadcaster)
+				{
+					Verbs:     util.NewStringSet("create", "update"),
+					Resources: util.NewStringSet("events"),
+				},
+			},
+		},
+		{
+			ObjectMeta: kapi.ObjectMeta{
+				Name: DeploymentControllerRoleName,
```
Comment (liggitt) [inline pkg/cmd/server/bootstrappolicy/policy.go]: @ironcladlou can you take a look at this list of permissions, and let me know if you see anything else the controller launched from RunDeploymentController() need? Note that the other RunDeploy*Controller functions are still using the highly-privileged client.

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 3]
```diff
@@ -213,6 +213,100 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
...
+				// DeploymentController.recorder (EventBroadcaster)
+				{
+					Verbs:     util.NewStringSet("create", "update"),
+					Resources: util.NewStringSet("events"),
+				},
+			},
+		},
+		{
+			ObjectMeta: kapi.ObjectMeta{
+				Name: ReplicationControllerRoleName,
```
Comment (liggitt) [inline pkg/cmd/server/bootstrappolicy/policy.go]: @ironcladlou can you also double-check these as permissions for the ReplicationController

File: pkg/cmd/server/bootstrappolicy/policy.go (hunk 1 above)
Comment (bparees) [inline pkg/cmd/server/bootstrappolicy/policy.go]: nothing else comes to mind for the BuildController.

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 4]
```diff
@@ -213,6 +213,100 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
...
+				},
+				// DeploymentControllerFactory.deploymentClient
+				{
+					Verbs:     util.NewStringSet("get", "update"),
+					Resources: util.NewStringSet("replicationcontrollers"),
+				},
+				// DeploymentController.podClient
+				{
+					Verbs:     util.NewStringSet("get", "create", "delete", "update"),
+					Resources: util.NewStringSet("pods"),
```
Comment (ironcladlou) [inline pkg/cmd/server/bootstrappolicy/policy.go]: `getDeployerPodsForFunc` uses a pod `list`.

File: pkg/cmd/server/bootstrappolicy/policy.go (hunk 3 above)
Comment (ironcladlou) [inline pkg/cmd/server/bootstrappolicy/policy.go]: Cursory scan looks correct, but I'm not familiar with that code so couldn't hurt to have somebody who is triple check.

File: pkg/cmd/server/bootstrappolicy/policy.go (hunk 4 above)
Comment (liggitt) [inline pkg/cmd/server/bootstrappolicy/policy.go]: added

**Comment by liggitt** [comment]
flake:
```
 'oc delete all --all' exited with status 1
```

**Comment by ironcladlou** [comment]
@liggitt 

>flake:
>'oc delete all --all' exited with status 1

I see:

    --- FAIL: TestBuildRunningPodDeleteController (30.62s)
	buildcontroller_test.go:332: Timed out waiting for watch: build updated to error

From: https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/2961/console

I mention this because the delete issues were supposed to be sorted out in master, and if they're not I need to know.

**Comment by liggitt** [comment]
I'm seeing it alternately fail on delete all and a random TestBuild* integration test. Neither recreatable locally

**Comment by smarterclayton** [comment]
A masterpiece of brevity and clarity.  Code Reviewers Weekly gives it 5 thumbs up

**Comment by smarterclayton** [comment]
Approved LGTM [merge]

**Pull request merged**